File: windows/src/Switcher3way.Core/SoftGates.cs
namespace Switcher3way.Core;

/// <summary>
/// Precision-first soft vetoes and letter-core trimming, shared by the auto and manual paths.
/// A faithful port of the macOS <c>LayoutDetector.passesSoftGates</c> + <c>NWayResolver.letterCore</c>.
/// A word enters detection only if it is a "real" word — not a single letter, an acronym, or code.
/// </summary>
public static class SoftGates
{
    /// <summary>
    /// True if <paramref name="typed"/> may be considered for conversion. Precision-first: on doubt,
    /// false. Callers pass the letter core (edge punctuation already trimmed).
    /// </summary>
    public static bool PassesSoftGates(string typed, bool capsLock)
    {
        if (typed.Length < 2) return false;                     // 1 letter (я/a/i/і): hopelessly ambiguous
        // digits/punctuation/URL/code/email — with one exception: an apostrophe or hyphen INSIDE the
        // word. "you're", "it's", "кто-то", "будь-ласка" are ordinary words of these languages and
        // already pass when typed in their own layout; vetoing them as code cost "You're", "That's"
        // and "Кто-то" their conversion in a field log. Edges are trimmed by LetterCore before this
        // runs, so a joiner can only be internal here — but check anyway.
        for (int i = 0; i < typed.Length; i++)
        {
            if (char.IsLetter(typed[i])) continue;
            if (!IsWordJoiner(typed[i]) || i == 0 || i == typed.Length - 1) return false;
        }
        // Under Caps Lock everything is UPPERCASE — not an acronym, not camelCase — so skip these two.
        if (!capsLock)
        {
            if (IsAllCaps(typed)) return false;                 // acronyms
            if (LooksLikeCodeIdentifier(typed)) return false;   // camelCase / mixed alphabets
        }
        return true;
    }

    /// <summary>
    /// The non-letters a word may carry inside it and still be a word: the apostrophe (straight or
    /// typographic) and the hyphen. Nothing else — a digit, a slash, an underscore is code.
    /// </summary>
    private static bool IsWordJoiner(char c) => c == '\'' || c == '\u2019' || c == '-';

    private static bool IsAllCaps(string s) =>
        s == s.ToUpperInvariant() && s != s.ToLowerInvariant();

    /// <summary>
    /// Looks like a code identifier: an internal capital (camelCase/PascalCase) or a mix of Latin
    /// and Cyrillic in one token → almost always code, not a word.
    /// </summary>
    private static bool LooksLikeCodeIdentifier(string s)
    {
        for (int i = 1; i < s.Length; i++)
            if (char.IsUpper(s[i])) return true;

        bool hasLatin = false, hasCyrillic = false;
        foreach (var ch in s)
        {
            int u = ch;
            if ((u >= 0x41 && u <= 0x5A) || (u >= 0x61 && u <= 0x7A)) hasLatin = true;
            else if (u >= 0x0400 && u <= 0x04FF) hasCyrillic = true;
        }
        return hasLatin && hasCyrillic;
    }

    /// <summary>The word's letter core: the string with leading/trailing non-letters trimmed.</summary>
    public static string LetterCore(string s)
    {
        int lo = 0, hi = s.Length;
        while (lo < hi && !char.IsLetter(s[lo])) lo++;
        while (hi > lo && !char.IsLetter(s[hi - 1])) hi--;
        return s.Substring(lo, hi - lo);
    }
}

File: Sources/Switcher3wCore/SoftGates.swift
import Foundation

/// The cheap vetoes applied before any dictionary lookup: we let a word into the detector only if
/// it is a "real" word, and not a single letter, an acronym, or code. Precision-first — on doubt,
/// false. Shared by the 2-way path (`LayoutDetector.decide`) and the N-way resolver.
public enum SoftGates {

    public static func passes(_ typed: String, capsLock: Bool) -> Bool {
        guard typed.count >= 2 else { return false }                  // 1 letter (я/a/i/і): hopelessly ambiguous between layouts
        guard typed.allSatisfy({ $0.isLetter }) else { return false } // digits/punctuation/URL/code/email
        // Under Caps Lock all text is UPPERCASE — this is NOT an acronym and NOT camelCase,
        // so these two vetoes are applied only when Caps Lock is off.
        if !capsLock {
            if isAllCaps(typed) { return false }                      // acronyms
            if looksLikeCodeIdentifier(typed) { return false }        // camelCase / mixed alphabets
        }
        return true
    }

    static func isAllCaps(_ s: String) -> Bool {
        s == s.uppercased() && s != s.lowercased()
    }

    /// Looks like a code identifier: an internal capital (camelCase/PascalCase)
    /// or a mix of Latin and Cyrillic in one token → almost always code, not a word.
    static func looksLikeCodeIdentifier(_ s: String) -> Bool {
        for (i, c) in s.enumerated() where i > 0 && c.isUppercase { return true }
        var hasLatin = false, hasCyrillic = false
        for u in s.unicodeScalars {
            switch u.value {
            case 0x41...0x5A, 0x61...0x7A: hasLatin = true
            case 0x0400...0x04FF: hasCyrillic = true
            default: break
            }
        }
        return hasLatin && hasCyrillic
    }

    /// The contiguous range of `chars` with leading/trailing characters that satisfy `drop` removed.
    static func coreRange(count: Int, drop: (Int) -> Bool) -> Range<Int> {
        var lo = 0, hi = count
        while lo < hi && drop(lo) { lo += 1 }
        while hi > lo && drop(hi - 1) { hi -= 1 }
        return lo..<hi
    }

    /// The word's letter core: the text with leading/trailing non-letters trimmed. Validation runs
    /// on the core so a trailing "!" or a leading "(" doesn't hide an otherwise-valid word, while
    /// the whole token is still re-rendered on output.
    public static func letterCore(_ chars: [Character]) -> String {
        String(chars[coreRange(count: chars.count) { !chars[$0].isLetter }])
    }

    public static func letterCore(_ s: String) -> String {
        letterCore(Array(s))
    }
}
